src/puppetlabs/puppetdb/command/dlo.clj

```diff
@@ -1,210 +1,173 @@
+  [stats command size]
+  (let [addnp (fn [v n] (+ (or v 0) n))]
+    (-> stats
+        (update-in ["all" :count] inc)
+        (update-in ["all" :size] addnp size)
+        (update-in [command :count] addnp 1)
+        (update-in [command :size] addnp size))))
+
+(defn- gather-stats [dlo]
+  "Gathers statistics about the filenames directly inside the dlo
+  directory (a Path).  Does not recurse.  e.g.:
+    {\"all\" {:count 104 :size 2073951}
+     \"replace facts\" {:count 34 :size 302116}
+     \"replace catalog\" {:count 19 :size 708182}
+     ...}"
+  (with-open [path-stream (Files/newDirectoryStream dlo)]
+    (reduce (fn [stats p]
+              ;; Assume the trailing .json here and in
+              ;; entry-cmd-err-filename below.
+              (if-let [cmeta (and (.endsWith p ".json")
+                                  (parse-cmd-filename (-> p .getFileName str)))]
+                (update-cmd-stats (:command cmeta) (Files/size p))
+                stats))
+            {"all" {:count 0 :size 0}}
+            path-stream)))
+
+(defn- cmd-gauges [dlo category]
+  "Adds gauges to the dlo for the given category (e.g. \"all\"
+  \"replace command\").  The gauges will pull their values from the
+  dlo atom's :stats."
+  (let [registry (:registry @dlo)
+        stat (fn [what] (get-in (:stats @dlo) what))]
+    {:count (gauge-fn registry ["puppetlabs.puppetdb.dlo" category "count"]
+                      #(stat [category :count]))
+     :size (gauge-fn registry ["puppetlabs.puppetdb.dlo" category "size"]
+                     #(stat [category :size]))}))
+
+(defn- note-new-discard! [dlo command size]
+  ;; Swap twice so that :stats will already have the values that the
+  ;; gauges are going to refer to.
+  (swap! dlo update :stats update-cmd-stats command size)
+  (swap! dlo assoc-in [:metrics command] (cmd-gauges dlo command)))
+
+(defn- write-event-info
+  "Writes a command attempt summary to out, including timestamp,
+  error information, and stacktrace."
+  [out what timestamp error trace]
+  (.write out (format "%s @ %s\n" what timestamp))
+  (.write out error)
+  (.newLine out)
+  (doseq [elt trace]
+    (.write out (str elt)) (.newLine out)))
+
+(defn- write-failure-metadata
+  "Given a (possibly empty) sequence of command attempts and an exception,
+  writes a summary of the failure to out."
+  [out attempts exception]
+  (dorun (map-indexed (fn [i {:keys [timestamp error trace]}]
+                        (write-event-info out
+                                          (str (when-not (zero? i) \newline)
+                                               "Attempt " (inc i))
+                                          timestamp error trace))
+                      attempts))
+  (.newLine out)
+  (write-event-info out
+                    (str (when-not (zero? (count attempts)) \newline)
+                         "Attempt " (inc (count attempts)))
+                    (timestamp)
+                    (str exception)
```

Comment:
Wait, so I'm confused -- while this won't indent the stacktrace, I think it may already print more or less exactly what you have above, and if we think it's better to trade RAM for a larger write (and simpler code), I we can just `(.write out (str/join "\n  " trace))`, or use printStackTrace for the output?